### PR TITLE
MAINT: inject external origination timestamp

### PR DESCRIPTION
**Pull request by chenqi0805**
### Description
For the purpose of tracking pipelineEndToEndLatency metric.
 
### Issues Resolved
Resolves #[Issue number to be closed when this PR is merged]
 
### Check List
- [ ] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [ ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by oeyh — APPROVED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java [hunk 1]
```diff
@@ -428,6 +428,7 @@ private <T> Record<Event> getRecord(ConsumerRecord<String, T> consumerRecord, in
         }
         eventMetadata.setAttribute("kafka_topic", topicName);
         eventMetadata.setAttribute("kafka_partition", String.valueOf(partition));
+        eventMetadata.setExternalOriginationTime(Instant.ofEpochMilli(consumerRecord.timestamp()));
```
[the remark this answers is not in the document]
Comment (chenqi0805) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java:431]: > Just out of curiosity: Is there the reverse encoding also taking place when writing to Kafka?
I am not expert on this but if the conversion to bytes happens, it probably happens under the hood in Kafka

[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java:431]: @KarstenSchnitter when a Kafka producer writes to kafka, Kafka automatically captures the time which can be retrieved at the time of consumption using the `consumerRecord.timestamp()`. We are currently not extracting any message attributes set by the producer.

**Review by kkondaka — APPROVED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (KarstenSchnitter) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java:431]: @kkondaka I was wondering, whether the producer set the timestamp manually to a certain value or whether the automatic timestamp by Kafka is used. The name "externalOriginationTime" would suggest to me, that the producer explicitly sets the timestamp to some value provided by the event. With your description, it sounds more like an "internalIngressTime" (or similar). I was confused by the naming.

[the remark this answers is not in the document]
Comment (KarstenSchnitter) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaCustomConsumer.java:431]: I am aware how Kafka works. I was just asking, if the timestamp could be taken from the event itself. For example OpenTelemetry logs can contain an [observed time](https://github.com/open-telemetry/opentelemetry-proto/blob/f16a3fdf35e3ad4c9b8aaef705e234ef8a2abfdf/opentelemetry/proto/logs/v1/logs.proto#L142-L157), which could have been used here. I am totally fine with the chosen approach.

**Pull request merged**